jsf-ri/src/main/java/com/sun/faces/component/search/SearchExpressionHandlerImpl.java

```diff
@@ -0,0 +1,541 @@
+            }
+
+            // check again the remainingExpression
+            SearchExpressionHandler handler = facesContext.getApplication().getSearchExpressionHandler();
+            return handler.isPassthroughExpression(searchExpressionContext, remainingExpression);
+        }
+
+        return false;
+    }
+
+    @Override
+    public boolean isValidExpression(SearchExpressionContext searchExpressionContext, String expression) {
+        if (expression != null) {
+            expression = expression.trim();
+        }
+        
+        if (expression == null || expression.isEmpty()) {
+            return true;
+        }
+
+        if (expression.contains(KEYWORD_PREFIX)) {
+            FacesContext facesContext = searchExpressionContext.getFacesContext();
+            SearchExpressionHandler handler = facesContext.getApplication().getSearchExpressionHandler();
+            
+            // absolute expression and keyword as first command -> try again from ViewRoot
+            char separatorChar = facesContext.getNamingContainerSeparatorChar();
+            if (expression.charAt(0) == separatorChar) {
+                expression = expression.substring(1);
+            }
+
+            String command = extractFirstCommand(facesContext, expression);
+
+            // check if there are remaining commands/id's after the first command
+            String remainingExpression = null;
+            if (command.length() < expression.length()) {
+                remainingExpression = expression.substring(command.length() + 1);
+            }
+
+            if (command.startsWith(KEYWORD_PREFIX)) {   
+                String keyword = command.substring(KEYWORD_PREFIX.length());
+
+                // resolver for keyword available?
+                SearchKeywordResolver keywordResolver = facesContext.getApplication().getSearchKeywordResolver();
+                if (!keywordResolver.isResolverForKeyword(searchExpressionContext, keyword)) {
+                    return false;
+                }
+
+                if (remainingExpression != null && !remainingExpression.trim().isEmpty()) {
+                    // there is remaingExpression avialable but the current keyword is leaf -> invalid
+                    if (keywordResolver.isLeaf(searchExpressionContext, keyword)) {
+                        return false;
+                    }
+
+                    return handler.isValidExpression(searchExpressionContext, remainingExpression);
+                }
+            }
+            else {
+                if (remainingExpression != null) {
+                    return handler.isValidExpression(searchExpressionContext, remainingExpression);
+                }
+            }
+        }
+
+        return true;
+    }
+
+    protected boolean isHintSet(SearchExpressionContext searchExpressionContext, SearchExpressionHint hint) {
+        if (searchExpressionContext.getExpressionHints() == null) {
```

Comment:
Javadoc implies it is never null. Make sure the actual impl returns emptySet() in such case.